### PR TITLE
[OC-1683] Always use perimeter when routing cards

### PR DESCRIPTION
**Pull request by vlo-rte**
Release notes : 

In Features section : 

OC-1683 : Always use perimeter when routing cards 

Signed-off-by: vlo-rte <valerie.longa@rte-france.com>

### CODE REVIEW

File: src/test/resources/perimeters/cypress.json [hunk 1]
```diff
@@ -1,18 +1,22 @@
...
-		{
-			"state": "messageStateWithNoAck",
-			"right": "ReceiveAndWrite"
-		},
-		{
-			"state": "messageStateWithOnlyWhenResponseDisabledForUserAck",
-			"right": "ReceiveAndWrite"
-		}
-	]
+  "id": "cypress",
```
Comment (AlexGuironnetRTE) [inline src/test/resources/perimeters/cypress.json:2]: I had the following message in the console:

> WARNING : state responseState is present in perimeter for process cypress but not in process definition

I'm guessing responseState should be removed from this file or added in the cypress bundle?

**Review by AlexGuironnetRTE — CHANGES_REQUESTED** [review]
Manual tests ok, unit tests / automatic tests are clear.
I have a question regarding "processStateKey" and I would suggest a few minor changes to the comments.
Other than that, shouldn't we add something to the release notes (or migration guide) to tell people what they should do if they were using user or groups recipients so far ?

File: services/core/cards-consultation/src/main/java/org/opfab/cards/consultation/model/ArchivedCardConsultationData.java [hunk 2]
```diff
@@ -96,4 +97,6 @@
     private String  representative;
     private PublisherTypeEnum representativeType;
     private Integer  secondsBeforeTimeSpanForReminder;
+
+    private String processStateKey;
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply services/core/cards-consultation/src/main/java/org/opfab/cards/consultation/model/ArchivedCardConsultationData.java:101]: In cards collection, processStateKey is an indexed field, so I think the performance is better with this field.

File: services/core/cards-consultation/src/main/java/org/opfab/cards/consultation/repositories/UserUtilitiesCommonToCardRepository.java [hunk 3]
```diff
@@ -67,37 +68,42 @@ default Criteria computeCriteriaForUser(CurrentUserWithPerimeters currentUserWit
...
+
+    1) If the card is sent to user1, the card is received and visible for user1 if he has the receive right for the
+       corresponding process/state (Receive or ReceiveAndWrite)
+    2) If the card is sent to GROUP1 (or ENTITY1), the card is received and visible for user if he :
+         - is member of GROUP1 (or ENTITY1)
+         - has the receive right for the corresponding process/state (Receive or ReceiveAndWrite)
+    3) If the card is sent to ENTITY1 and GROUP1, the card is received and visible for user if he :
+         - is member of ENTITY1
+         - is member of GROUP1
+         - has the receive right for the corresponding process/state (Receive or ReceiveAndWrite)
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply services/core/cards-consultation/src/main/java/org/opfab/cards/consultation/repositories/UserUtilitiesCommonToCardRepository.java:89]: Done.

File: services/core/cards-consultation/src/main/java/org/opfab/cards/consultation/services/CardSubscription.java [hunk 4]
```diff
@@ -298,23 +299,32 @@ public boolean checkIfUserMustBeNotifiedForThisProcessState(String process, Stri
...
+    /** Rules for receiving cards :
+    1) If the card is sent to user1, the card is received and visible for user1 if he has the receive right for the
+       corresponding process/state (Receive or ReceiveAndWrite)
+    2) If the card is sent to GROUP1 (or ENTITY1), the card is received and visible for user if he :
+         - is member of GROUP1 (or ENTITY1)
+         - has the receive right for the corresponding process/state (Receive or ReceiveAndWrite)
+    3) If the card is sent to ENTITY1 and GROUP1, the card is received and visible for user if he :
+         - is member of ENTITY1
+         - is member of GROUP1
+         - has the receive right for the corresponding process/state (Receive or ReceiveAndWrite)
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply services/core/cards-consultation/src/main/java/org/opfab/cards/consultation/services/CardSubscription.java:324]: Done.

File: src/test/resources/perimeters/cypress.json (hunk 1 above)
Comment (vlo-rte) [reply src/test/resources/perimeters/cypress.json:2]: Removed from this file.

**Comment by vlo-rte** [comment]
> Manual tests ok, unit tests / automatic tests are clear.
> I have a question regarding "processStateKey" and I would suggest a few minor changes to the comments.
> Other than that, shouldn't we add something to the release notes (or migration guide) to tell people what they should do if they were using user or groups recipients so far ?

You're right. I suggest to add the following : 
+
IMPORTANT: From now, whatever the type(s) of recipient(s), the perimeters are always checked. So if you are using users directly or groups as card recipients, you will now need to add receive rights (Receive or ReceiveAndWrite) for the corresponding process/state of the card.
+

**Review by AlexGuironnetRTE — APPROVED** [review]

**Pull request merged**